### ISSUE TITLE
esm example not working

### ISSUE THREAD

**Issue by cherviakovtaskworld** [oclif/oclif#1364]
**Describe the bug**
Trying to generate esm hello world example and make it work, but getting command hello (or hello:world) not found error

**To Reproduce**
1. Follow instructions in readme/guide to generate new plugin
2. Choose ESM and pnpm (btw result folder also includes both yarn and pnpm lock files in the end)
3. After all done try to execute node bin/dev.js hello
4. See error message saying that command not found

**Expected behavior**
Should be running hello command, same as in commonjs case

**Screenshots**
N/A

**Environment (please complete the following information):**
 - OS Archlinux
 - Shell/terminal zsh 5.9

**Additional context**
Please add full list of all packages need to be installed in your guides/README (I mean it is nice to have generate and learn from example but sometimes need to be able to gather something working manually)
Would be nice to have somewhere explanation of oclif workflow for better understanding how it reading config, search for commands, parsing arguments and finally executing it
Getting some mysterious TypeError logs during tries of executing esm example
In my working code also got errors related to unknown file extention, but same as TypeError above not anything which can help understand what is wrong.

**Comment by mdonnalley** [comment]
@cherviakovtaskworld I'm not seeing that issue on latest (4.7.7). What version are you using?

Good catch on the yarn.lock still existing when you select pnpm. That should be an easy fix for.

```
❯ oclif --version
oclif/4.7.7 darwin-arm64 node-v20.11.0

❯ oclif generate gh1364 --module-type ESM --package-manager pnpm --yes
Generating gh1364 in /Users/mdonnalley/code/gh1364
? Select a module type ESM
? NPM package name gh1364
? Command bin name the CLI will export gh1364
? Description oclif example Hello World CLI (ESM)
? Author Mike Donnalley @mdonnalley
? License MIT
? Who is the GitHub owner of repository (https://github.com/OWNER/repo) mdonnalley
? What is the GitHub name of repository (https://github.com/owner/REPO) gh1364
? Select a package manager pnpm
pnpm install
Progress: resolved 1, reused 0, downloaded 0, added 0
Progress: resolved 86, reused 69, downloaded 13, added 0
Progress: resolved 222, reused 180, downloaded 38, added 0
Progress: resolved 289, reused 198, downloaded 78, added 0
Progress: resolved 420, reused 289, downloaded 119, added 0
Progress: resolved 530, reused 357, downloaded 155, added 0
Packages: +611
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 611, reused 402, downloaded 208, added 0
Progress: resolved 611, reused 402, downloaded 208, added 399
Progress: resolved 611, reused 402, downloaded 209, added 611, done

> gh1364@0.4.28 prepare /Users/mdonnalley/code/gh1364
> pnpm run build


> gh1364@0.4.28 build /Users/mdonnalley/code/gh1364
> shx rm -rf dist && tsc -b


dependencies:
+ @oclif/core 3.26.2
+ @oclif/plugin-help 6.0.20
+ @oclif/plugin-plugins 5.0.7

devDependencies:
+ @oclif/prettier-config 0.2.1
+ @oclif/test 3.2.8
+ @types/chai 4.3.14
+ @types/mocha 10.0.6
+ @types/node 18.19.31 (20.12.6 is available)
+ chai 4.4.1 (5.1.0 is available)
+ eslint 8.57.0 (9.0.0 is available)
+ eslint-config-oclif 5.1.1
+ eslint-config-oclif-typescript 3.1.4
+ eslint-config-prettier 9.1.0
+ mocha 10.4.0
+ oclif 4.7.7
+ shx 0.3.4
+ ts-node 10.9.2
+ typescript 5.4.4

Done in 10.3s
/Users/mdonnalley/code/gh1364/node_modules/.bin/oclif readme
replacing <!-- usage --> in README.md
replacing <!-- commands --> in README.md
replacing <!-- toc --> in README.md

Created gh1364

❯ cd gh1364

❯ bin/dev.js hello cherviakovtaskworld --from mdonnalley
hello cherviakovtaskworld from mdonnalley! (./src/commands/hello/index.ts)
```

**Comment by mdonnalley** [comment]
@cherviakovtaskworld I just realized that you're running `node bin/dev.js` instead of just `bin/dev.js`

Using `node` is going to override the hashbang in `bin/dev.js`, which uses the `ts-node/esm` loader, which is what allows node to run typescript ESM files. See https://github.com/oclif/hello-world-esm/blob/main/bin/dev.js#L1

So you need to do one of the following in order to successfully run `bin/dev.js` in an ESM project
- `bin/dev.js`
- `node --loader ts-node/esm bin/dev.js`
- `ts-node bin/dev.js`

**Comment by cherviakovtaskworld** [comment]
Thanks for quick response. Didn't know that hashbang can be overriden, let me try again.

**Comment by cherviakovtaskworld** [comment]
Indeed working, then all good.

**Comment by cherviakovtaskworld** [comment]
Hmm running some experiments trying to understand how such scripts executes, surprisingly code of bin/dev can be javascript but not typescript, while if I import typescript files, they processed and working!
Then requirements for bin/dev to have extention .js, need to type: module in package.json, tsconfig.json - lots of ways how you can run and configure it and good deal of various errors you will get.

I think complexity here is quite high, you need to be very specific on all setup and ways to run it.
